x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -159,6 +165,69 @@ public void updateProfileData(UpdateProfileDataRequest request, ActionListener<A
         );
     }
 
+    public void searchProfile(SearchProfilesRequest request, ActionListener<SearchProfilesResponse> listener) {
+        tryFreezeAndCheckIndex(listener).ifPresent(frozenProfileIndex -> {
+            final BoolQueryBuilder query = QueryBuilders.boolQuery().must(QueryBuilders.termQuery("user_profile.enabled", true));
+            if (Strings.hasText(request.getQuery())) {
+                query.must(
+                    QueryBuilders.multiMatchQuery(
+                        request.getQuery(),
+                        "user_profile.user.username",
+                        "user_profile.user.username._2gram",
+                        "user_profile.user.username._3gram",
+                        "user_profile.user.full_name",
+                        "user_profile.user.full_name._2gram",
+                        "user_profile.user.full_name._3gram",
+                        "user_profile.user.display_name",
+                        "user_profile.user.display_name._2gram",
+                        "user_profile.user.display_name._3gram"
+                    ).type(MultiMatchQueryBuilder.Type.BOOL_PREFIX)
+                );
+            }
+            final SearchRequest searchRequest = client.prepareSearch(SECURITY_PROFILE_ALIAS)
+                .setQuery(query)
+                .setSize(request.getSize())
+                .addSort("_score", SortOrder.DESC)
+                .addSort("user_profile.uid", SortOrder.ASC)
+                .request();
+
+            frozenProfileIndex.checkIndexVersionThenExecute(
+                listener::onFailure,
+                () -> executeAsyncWithOrigin(
+                    client,
+                    SECURITY_ORIGIN,
```

Comment:
I am not sure about the value of adding a new origin unless we also add a different internal user that is not `_xpack_security`. Maybe we should do it. But I think a separate PR is OK since it is only just for this single API?

Reply:
Yes, a new origin and a new internal user. Of course a separate PR.